New hoomd example

atravitz opened the pull request:
For a while we've been needing to update some of these examples. Here's a proposed update to the HOOMD-blue example. The most notable change is the addition of a walk-through readme.

mikemhenry commented:
I really like this example!

mikemhenry commented on projects/flow.hoomd.lj/src/README.md:
```diff
@@ -0,0 +1,20 @@
+    - If you look in these directories, you'll see `signac_statepoint.json`. This is a json file that contains the statepoint parameters for that job.
+    - NOTE: The job's `id` is generated specifically for the dict containing the statepoint parameters, so do not directly change the directory name or `signac_statepoint.json` directly.
+
+2. All of the operations we will be performing on our data is stored in `project.py`.
```
Maybe add some quick definition of what an "operation" is?

mikemhenry commented on projects/flow.hoomd.lj/src/README.md:
```diff
@@ -0,0 +1,20 @@
+
+4. For more information on how you can analyze this data, take a look at the `visualize_data.ipynb` jupyter notebook in this directory.
+
+NOTE: If you want to run this tutorial from scratch, just `rm -rf workspace/`
```
```suggestion
NOTE: If you want to run this tutorial from scratch, just run `rm -rf workspace/` to delete the workspace.
```

mikemhenry commented on projects/flow.hoomd.lj/src/init.py:
```diff
@@ -3,60 +3,38 @@
-                    # barostat coupling constant
-                    tauP=1.0)
-            project.open_job(statepoint).init()
+    for p in np.linspace(0.5, 5.0, 10):
```
Maybe add a code comment here, something like "sweep over 10 different pressures from 0.5 to 5.0 at 0.5 pressure increments " not sure on the exact wording but people might not know exactly what np.linspace does

csadorf commented:
@atravitz I believe this is ready for another round of review? If yes, could you request that?

Besides that, it appears that there are some build issues, any idea what might be going on?

bdice reviewed (changes requested):
@atravitz The style of the README has to follow a particular format. The continuous integration we added (and which used to be enabled long ago) parses the README and runs the commands it finds in code blocks beneath the heading `# Usage`. The updated text is definitely useful but the list formatting will need to look more like the previous.

Also should `status.txt` be removed (and added to `.gitignore`)?

atravitz commented:
> @atravitz The style of the README has to follow a particular format. The continuous integration I added in #17 (and which used to be enabled long ago) parses the README and runs the commands it finds in code blocks beneath the heading `# Usage`. The updated text is definitely useful but the list formatting will need to look more like the previous. If you want help or an example, let me know.
> 
> Also should `status.txt` be removed (and added to `.gitignore`)?

Should we have `print_status.sh` in there at all?

bdice commented:
> Would breaking the commands into code blocks be sufficient? Should we have print_status.sh in there at all?

@atravitz I think so. You can refer to the other examples. The CI will run `python flow-test.py flow.hoomd.lj` from the `projects/` directory, and you can also run that locally to test it.

I don't think we need `print_status.sh`, you can remove it along with `status.txt`.

atravitz commented:
> > Would breaking the commands into code blocks be sufficient? Should we have print_status.sh in there at all?
> 
> @atravitz I think so. You can refer to the other examples. The CI will run `python flow-test.py flow.hoomd.lj` from the `projects/` directory, and you can also run that locally to test it.
> 
> I don't think we need `print_status.sh`, you can remove it along with `status.txt`.

@bdice let me know if this looks good to you

bdice commented on projects/flow.hoomd.lj/README.md:
````diff
@@ -1,53 +1,66 @@
-```
-python src/project.py run
-```
+4. For more information on how you can analyze this data, imply execute ``jupyter notebook`` within the project's root directory and open the `src/analysis.ipynb` notebook.
````
This was supposed to say "simply" instead of "imply" but I think removing the word is better.
```suggestion
4. For more examples of how you can analyze this data, execute ``jupyter notebook`` within the project's root directory and open the `src/analysis.ipynb` notebook.
```

bdice commented on projects/flow.hoomd.lj/README.md:
```diff
@@ -1,53 +1,66 @@
+
+- This creates the `workspace/`, which holds all of our `jobs`. Each `job` has its own directory, named by the job's unique `id` (something like `87c7fccdea3531da704bbae95e95e914`).
+- If you look in these directories, you'll see `signac_statepoint.json`. This is a json file that contains the statepoint parameters for that job.
+- NOTE: The job's `id` is generated specifically for the dict containing the statepoint parameters, so do not directly change the directory name or `signac_statepoint.json` directly.
```
```suggestion
- NOTE: The job's `id` is generated specifically for the dict containing the statepoint parameters, so do not edit the directory name or `signac_statepoint.json`.
```

bdice commented on projects/flow.hoomd.lj/README.md:
```diff
@@ -1,53 +1,66 @@
+- If you look in these directories, you'll see `signac_statepoint.json`. This is a json file that contains the statepoint parameters for that job.
+- NOTE: The job's `id` is generated specifically for the dict containing the statepoint parameters, so do not directly change the directory name or `signac_statepoint.json` directly.
+
+2. All of the operations we will be performing on our data is stored in `project.py`. An operation is a function with `job` as its only argument that signac-flow recognizes as a part of your workflow.
```
```suggestion
2. All of the operations we will be performing on our data are defined in `project.py`. An operation is a function with `job` as its only argument that signac-flow recognizes as a part of your workflow.
```

bdice commented on projects/flow.hoomd.lj/README.md:
```diff
@@ -1,53 +1,66 @@
+- NOTE: The job's `id` is generated specifically for the dict containing the statepoint parameters, so do not directly change the directory name or `signac_statepoint.json` directly.
+
+2. All of the operations we will be performing on our data is stored in `project.py`. An operation is a function with `job` as its only argument that signac-flow recognizes as a part of your workflow.
+    - You can tell which methods are operations because they will have the `@MyProject.operation` decorator, which tells signac that this method is to be treated as an operation.
```
```suggestion
    - You can tell which methods are operations because they will have the `@MyProject.operation` decorator, which tells signac-flow that this method is to be treated as an operation associated with `MyProject`.
```

bdice commented on projects/flow.hoomd.lj/README.md:
```diff
@@ -1,53 +1,66 @@
+
+2. All of the operations we will be performing on our data is stored in `project.py`. An operation is a function with `job` as its only argument that signac-flow recognizes as a part of your workflow.
+    - You can tell which methods are operations because they will have the `@MyProject.operation` decorator, which tells signac that this method is to be treated as an operation.
+    - `operation`s typically have pre- and post- conditions. This is how signac knows when the operation should be run. For example, the `initialize()` function has no pre-condition because it must be run first, but has `@MyProject.post.isfile('init.gsd')` as a post-condition. This post-condition means that the operation will *not* be run if `init.gsd` exists.
```
```suggestion
    - `operation`s typically have pre- and post- conditions. This is how signac-flow knows when the operation should be run. For example, the `initialize()` function has no pre-condition because it must be run first, but has `@MyProject.post.isfile('init.gsd')` as a post-condition. This post-condition means that the operation will *not* be run if `init.gsd` exists in the job directory.
```

bdice commented on projects/flow.hoomd.lj/README.md:
```diff
@@ -1,53 +1,66 @@
+    - You can tell which methods are operations because they will have the `@MyProject.operation` decorator, which tells signac that this method is to be treated as an operation.
+    - `operation`s typically have pre- and post- conditions. This is how signac knows when the operation should be run. For example, the `initialize()` function has no pre-condition because it must be run first, but has `@MyProject.post.isfile('init.gsd')` as a post-condition. This post-condition means that the operation will *not* be run if `init.gsd` exists.
+
+3. Now run let's run the operations:
```
```suggestion
3. Now let's run the operations:
```

bdice commented on projects/flow.hoomd.lj/README.md:
```diff
@@ -1,53 +1,66 @@
 
-The project's workflow and data space operations are defined within the `src/project.py` module.
-We can check the project's status, e.g., with:
+    This will run just the `initialize()` operations for *all* eligible jobs (which in this case is all of the jobs).
```
```suggestion
    This will run just the `initialize()` operation for *all* eligible jobs (which in this case is all of the jobs).
```

bdice commented on projects/flow.hoomd.lj/README.md:
```diff
@@ -1,53 +1,66 @@
-The project's workflow and data space operations are defined within the `src/project.py` module.
-We can check the project's status, e.g., with:
+    This will run just the `initialize()` operations for *all* eligible jobs (which in this case is all of the jobs).
+    - Run `python project.py status -d` again, and you'll see that now all the jobs are eligible for `estimate` and `sample`. You can also run `python project.py status -d -p p`. The ``-p`` argument specifies which parameters should be shown in the status view, and we pass in `p` to see which statepoint corresponds to which pressure.
```
```suggestion
    - Run `python project.py status -d` again, and you'll see that now all the jobs are eligible for the operations `estimate` and `sample`. You can also run `python project.py status -d -p p`. The ``-p`` argument specifies which parameters should be shown in the status view, and we pass in `p` to see which statepoint corresponds to which pressure.
```

bdice commented on projects/flow.hoomd.lj/src/project.py:
```diff
@@ -49,24 +42,15 @@ def initialize(job):
-
-@MyProject.operation
-@MyProject.pre(initialized)
+@MyProject.pre.isfile("init.gsd")
```
Would it be better to use `after`?
```suggestion
@MyProject.pre.after(initialize)
```

atravitz replied:
Personally I like using `isfile` here because it seems like the most explicit way of showing how pre/post conditions work.

bdice reviewed (changes requested):
Thanks for the revisions @atravitz. I have some suggestions below.

bdice commented on projects/flow.hoomd.lj/README.md:
```diff
@@ -1,53 +1,66 @@
 Other modules:
 
-  * `src/environment.py` - Custom **environment** definitions
+  * ``environment.py`` - Custom **environment** definitions
```
This doesn't exist anymore, right? Remove if not.

atravitz replied:
```suggestion

```

bdice reviewed (approved):
Looks great! Thanks @atravitz.

bdice commented:
I'm going to go ahead and merge this. The PR been stalled for a long time and it is a significant improvement over the existing tutorial.
